README.md

<p align="center">
<img src="docs/assets/banana-reporter-logo.png" alt="Banana Reporter Logo" height="70px"/>
</p>
<h1 align="center">Banana Reporter</h1>
<p align="center">Create a report in CSV, JSON and JSONL from multiple sources (GitLab, GitHub, Git local repository and todo.txt)</p>

<p align="center">
  <a href="https://oclif.io"><img src="https://img.shields.io/badge/cli-oclif-brightgreen.svg" alt="oclif"/></a>
  <a href="https://npmjs.org/package/bananareporter"><img src="https://img.shields.io/npm/v/bananareporter.svg" alt="Version"/></a>
  <a href="https://github.com/nya1/bananareporter/actions/workflows/test.yml"><img src="https://github.com/nya1/bananareporter/actions/workflows/test.yml/badge.svg" alt="Github"/></a>
  <a href="https://npmjs.org/package/bananareporter"><img src="https://img.shields.io/npm/dt/bananareporter.svg" alt="Total Downloads"/></a>
  <a href="https://github.com/nya1/bananareporter/blob/main/package.json"><img src="https://img.shields.io/npm/l/bananareporter.svg" alt="License"/></a>
</p>

<!-- toc -->
* [Features](#features)
* [Quickstart](#quickstart)
* [Usage](#usage)
* [Commands](#commands)
<!-- tocstop -->

# Features

- **Easy to use**: provide a `--from` and `--to` date range and a config file, all data will be fetched automatically
- **Multiple output formats**: output file can be a JSON, JSONL or CSV (via `--format`) _more coming soon_
- **Configurable**: data to be imported can be filtered using the configuration file and each option can be overridden per source
- **Sources Supported**
  - [GitLab](./docs/sources/gitlab.md): commits
  - [GitHub](./docs/sources/github.md): commits
  - [Git CLI (local repository)](./docs/sources/git-cli.md): commits
  - todo.txt: tasks
  - **[Request a new source](https://github.com/nya1/bananareporter/issues/new?assignees=&labels=enhancement&template=new-source-request.md&title=)**

## Use cases

- create a PDF work report based on commits and tasks
- get insights on your activity 

# Quickstart

0. Install banana reporter CLI `npm i -g bananareporter`

1. Create a config file `bananareporter.yaml` with the sources that you want to fetch

```yaml
sources:
  - type: 'gitlab'
    committerUsername: usernameOnGitlab
    token: glpat-personalaccesstoken
    # filters:
    #   - on: '$project.path_with_namespace'
    #     regex: '(namespace|anotherone)'
  - type: 'github'
    committerUsername: usernameOnGithub
    # optional, needed for commits on private repositories
    token: personalaccesstoken
    # filters:
    #   - on: 'repository.full_name'
    #     regex: '(namespace|anotherone)'
  # - type: 'todo.txt'
  #   file: './todo.txt'
```

2. Run the reporter with a date range (ISO8601 date)

```sh
bananareporter --from 2023-01-01 --to 2023-03-01 -c bananareporter.yaml
```

In the current directory you will find the output as `bananareporter_$FROM__$TO.json` (can be changed with `--out`)

Example of output (json) with gitlab and github sources:

```json
[
  {
    "id": "c12ba180bfecf45fcdcc40d6104d1f1b7ad409dc",
    "date": "2023-01-13T07:51:21.730Z",
    "username": "johndoe",
    "description": "chore: update changelog and swagger branch:work git:aa33b04",
    "projectId": "3318214",
    "projectName": "awesome-frontend",
    "type": "gitlab"
  },
  {
    "id": "6e1b66a1dea89e957d8c44943f942be4874c0641",
    "date": "2023-01-14T10:50:10.230Z",
    "username": "johndoe",
    "description": "refactor: compare date function branch:work git:ia1f241",
    "projectId": "928544",
    "projectName": "awesome-backend",
    "type": "github"
  }
]
```


# Usage
<!-- usage -->
```sh-session
$ npm install -g bananareporter
$ bananareporter COMMAND
running command...
$ bananareporter (--version)
bananareporter/0.3.0 linux-x64 node-v16.19.0
$ bananareporter --help [COMMAND]
USAGE
  $ bananareporter COMMAND
...
```
<!-- usagestop -->
# Commands
<!-- commands -->
* [`bananareporter help [COMMANDS]`](#bananareporter-help-commands)
* [`bananareporter run`](#bananareporter-run)

## `bananareporter help [COMMANDS]`

Display help for bananareporter.

```
USAGE
  $ bananareporter help [COMMANDS] [-n]

ARGUMENTS
  COMMANDS  Command to show help for.

FLAGS
  -n, --nested-commands  Include all nested commands in the output.

DESCRIPTION
  Display help for bananareporter.
```

_See code: [@oclif/plugin-help](https://github.com/oclif/plugin-help/blob/v5.2.8/src/commands/help.ts)_

## `bananareporter run`

Run report

```
USAGE
  $ bananareporter run -o <value> --format json|jsonl|csv [-c <value>] [--to <value> --from <value>]
    [--delay <value>] [--include-raw-object]

FLAGS
  -c, --config=<value>  config file location, by default ~/.config/bananareporter/config.yaml
  -o, --out=<value>     (required) [default: ./bananareporter.json] file path to save the output
  --delay=<value>       [default: 300] global delay in millisecons between http requests
  --format=<option>     (required) [default: json] output file format
                        <options: json|jsonl|csv>
  --from=2023-03-01     from date (ISO8601)
  --include-raw-object  include raw object in json/jsonl reporter output
  --to=2023-03-31       to date (ISO8601)

DESCRIPTION
  Run report

EXAMPLES
  $ banana-reporter run --from 2023-01-01 --to 2023-01-31
  report with 138 entries saved to ./bananareporter.json
```

_See code: [dist/commands/run/index.ts](https://github.com/nya1/bananareporter/blob/v0.3.0/dist/commands/run/index.ts)_
<!-- commandsstop -->


CHANGELOG.md
## [0.4.3](https://github.com/nya1/bananareporter/compare/0.4.2...0.4.3) (2023-06-17)


### Bug Fixes

* **git-cli:** commit message escaping ([c01a3c3](https://github.com/nya1/bananareporter/commit/c01a3c3e4190bf678d97734010af2aeadc435a90))



## [0.4.2](https://github.com/nya1/bananareporter/compare/0.4.1...0.4.2) (2023-06-17)


### Bug Fixes

* readme typo ([7567325](https://github.com/nya1/bananareporter/commit/7567325b08c1783c483a6a86e81febd09599fe4c))



## [0.4.1](https://github.com/nya1/bananareporter/compare/0.4.0...0.4.1) (2023-06-17)



# [0.4.0](https://github.com/nya1/bananareporter/compare/0.3.3...0.4.0) (2023-06-17)


### Features

* add git-cli source type to load local repositories ([#24](https://github.com/nya1/bananareporter/issues/24)) ([f8ae15b](https://github.com/nya1/bananareporter/commit/f8ae15bcc3cf1b45ded34aefdc8276ffc56ef99a))



## [0.3.3](https://github.com/nya1/bananareporter/compare/0.3.2...0.3.3) (2023-03-26)


### Bug Fixes

* **deps:** bump papaparse from 5.4.0 to 5.4.1 ([3ad24d2](https://github.com/nya1/bananareporter/commit/3ad24d27c4e7d73000bebb95d4bea7c3795817d8))



## [0.3.2](https://github.com/nya1/bananareporter/compare/0.3.1...0.3.2) (2023-03-19)


### Bug Fixes

* **deps:** bump @oclif/core from 2.6.2 to 2.6.4 ([#16](https://github.com/nya1/bananareporter/issues/16)) ([33b7460](https://github.com/nya1/bananareporter/commit/33b7460aac4599670d4feb2eaedd6f18febf0d43))



## [0.3.1](https://github.com/nya1/bananareporter/compare/0.3.0...0.3.1) (2023-03-18)


### Bug Fixes

* remove duplicated entries ([2458859](https://github.com/nya1/bananareporter/commit/2458859d2c0003056dd3c6a9f6bfd90ed56af81d))



# [0.3.0](https://github.com/nya1/bananareporter/compare/0.2.1...0.3.0) (2023-03-18)


### Bug Fixes

* add from - to date in json output filename ([8aa7e9c](https://github.com/nya1/bananareporter/commit/8aa7e9c5b78289edd7265551851ae58806578a4d))
* skip duplicated commits ([2468abf](https://github.com/nya1/bananareporter/commit/2468abf3713da35a3a56fbeddf73fc766f978223))


### Features

* accept directories in --out flag ([59df613](https://github.com/nya1/bananareporter/commit/59df613d29bec5597c23212f4d42fd62cbbdea0b))
* add "id" on report object ([0b86c70](https://github.com/nya1/bananareporter/commit/0b86c70c219490f8b3b62e8af9678b4049d79aa8))



## [0.2.1](https://github.com/nya1/bananareporter/compare/0.2.0...0.2.1) (2023-03-12)


### Bug Fixes

* **deps:** bump @oclif/plugin-help from 5.2.7 to 5.2.8 ([#11](https://github.com/nya1/bananareporter/issues/11)) ([168af44](https://github.com/nya1/bananareporter/commit/168af44762a602328ea559a2c3552cb58899d977))
* **deps:** bump zod from 3.21.0 to 3.21.4 ([#10](https://github.com/nya1/bananareporter/issues/10)) ([548fcfc](https://github.com/nya1/bananareporter/commit/548fcfc30e0ef521af24fcc2d985708c882607a3))
* improve validation errors ([50e1a9e](https://github.com/nya1/bananareporter/commit/50e1a9e91620c9163303446146040d65883821f9))



# [0.2.0](https://github.com/nya1/bananareporter/compare/0.1.2...0.2.0) (2023-03-11)


### Features

* sort report by date ASC ([#9](https://github.com/nya1/bananareporter/issues/9)) ([bfdfc58](https://github.com/nya1/bananareporter/commit/bfdfc5815d0b0325818213a7ce8ec1cefd56f858))



## [0.1.2](https://github.com/nya1/bananareporter/compare/0.1.1...0.1.2) (2023-03-11)


### Bug Fixes

* README.md typo ([2cbad01](https://github.com/nya1/bananareporter/commit/2cbad015db7dddb985bb820bb85c512f7bc15f6d))



## [0.1.1](https://github.com/nya1/bananareporter/compare/0.1.0...0.1.1) (2023-03-11)


### Bug Fixes

* **gitlab:** empty commit title ([e5a6fd4](https://github.com/nya1/bananareporter/commit/e5a6fd43aef773efc58b894354b2ac4cd4628c7c))



# [0.1.0](https://github.com/nya1/bananareporter/compare/c86d6ad391f0353f2d4b9fbef4a5a823005fb015...0.1.0) (2023-03-04)


### Features

* add github and todo.txt integrations ([#7](https://github.com/nya1/bananareporter/issues/7)) ([c86d6ad](https://github.com/nya1/bananareporter/commit/c86d6ad391f0353f2d4b9fbef4a5a823005fb015))





src/commands/run/index.ts
import {Command, Flags} from '@oclif/core'
import * as yaml from 'js-yaml'
import {readFileSync, statSync, writeFileSync} from 'node:fs'
import path = require('node:path');
import dayjs = require('dayjs');
import {BananaOutputFileFormat, loadAndValidateConfig} from '../../core'
import {CommonBananaReporterObj, SourceType} from '../../integrations/base'
import {GitlabIntegration} from '../../integrations/gitlab'
import {unparse as jsonToCsv} from 'papaparse'
import {omit} from 'lodash'
import {logger} from '../../util/logger'
import {GithubIntegration} from '../../integrations/github'
import {TodoTxtIntegration} from '../../integrations/todotxt'
import {z} from 'zod'
import {resolve} from 'node:path'
import {GitCliIntegration} from '../../integrations/git-cli'

export default class Run extends Command {
  static description = 'Run report'

  static dateStringValidation(input: string): string {
    if (!dayjs(input).isValid()) {
      throw new Error(`provided date "${input}" is not valid, make sure it's valid ISO8601 string`)
    }

    return input
  }

  static examples = [
    `$ banana-reporter run --from 2023-01-01 --to 2023-01-31
report with 138 entries saved to ./bananareporter.json
`,
  ]

  static flags = {
    config: Flags.string({
      char: 'c',
      description: 'config file location, by default ~/.config/bananareporter/config.yaml',
      required: false,
      env: 'BANANAREPORTER_CONFIG_PATH',
    }),
    from: Flags.string({
      description: 'from date (ISO8601)',
      helpValue: `${dayjs().startOf('month').format('YYYY-MM-DD')}`,
      required: false,
    }),
    to: Flags.string({
      description: 'to date (ISO8601)',
      helpValue: `${dayjs().endOf('month').format('YYYY-MM-DD')}`,
      required: false,
      dependsOn: ['from'],
    }),
    out: Flags.file({
      char: 'o',
      aliases: ['output'],
      description: 'file path or directory to save the output',
      required: true,
      default: './bananareporter_$FROM__$TO.json',
    }),
    format: Flags.string({
      description: 'output file format',
      options: Object.values(BananaOutputFileFormat.Values),
      required: true,
      default: BananaOutputFileFormat.Enum.json,
    }),
    delay: Flags.integer({
      description: 'global delay in millisecons between http requests',
      default: 300,
    }),
    'include-raw-object': Flags.boolean({
      description: 'include raw object in json/jsonl reporter output',
      default: false,
    }),
    'disable-welcome-emoji': Flags.boolean({
      description: 'disable banana welcome emoji',
      default: false,
      hidden: true,
    }),
  }

  async run(): Promise<void> {
    // this.log('🍌 Banana Reporter')
    const {flags} = await this.parse(Run)

    logger.debug('run CLI flags', flags)

    const configFileLocation = flags.config || path.join(this.config.configDir, 'config.yaml')

    logger.debug(`run configFileLocation: ${configFileLocation}`)

    const configContent = yaml.load(readFileSync(configFileLocation).toString()) as Record<string, unknown>
    logger.debug('run configContent', configContent)

    const validatedConfig = loadAndValidateConfig({
      ...configContent,
      configFileLocation,
      ...omit(flags, ['include-raw-object', 'config']),
      includeRawObject: flags['include-raw-object'],
    })
    logger.debug('run validatedConfig', validatedConfig)

    // TODO the sources with different hosts / types can be called in parallel

    let reportList: CommonBananaReporterObj[] = []
    for (const [i, s] of validatedConfig.sources.entries()) {
      logger.debug(`run source loop ${i}`, s)
      this.log(`working on ${s.type} (${i + 1}) source`)
      let res: CommonBananaReporterObj[] = []
      switch (s.type) {
      case SourceType.Enum.gitlab: {
        const integrationLoaded = new GitlabIntegration(s, validatedConfig)
        // eslint-disable-next-line no-await-in-loop
        res = await integrationLoaded.fetchData()

        break
      }

      case SourceType.Enum.github: {
        const integrationLoaded = new GithubIntegration(s, validatedConfig)
        // eslint-disable-next-line no-await-in-loop
        res = await integrationLoaded.fetchData()

        break
      }

      case SourceType.Enum['git-cli']: {
        const integrationLoaded = new GitCliIntegration(s, validatedConfig)
        // eslint-disable-next-line no-await-in-loop
        res = await integrationLoaded.fetchData()

        break
      }

      case SourceType.Enum['todo.txt']: {
        const integrationLoaded = new TodoTxtIntegration(s, validatedConfig)
        // eslint-disable-next-line no-await-in-loop
        res = await integrationLoaded.fetchData()

        break
      }

      default: {
        this.error(`unknown source provided ${s.type}`)
      }
      }

      logger.debug(`run source loop integration res length ${res.length}`)

      // eslint-disable-next-line unicorn/prefer-spread
      reportList = reportList.concat(res)
    }

    // remove any duplicates (based on "id")
    reportList = [...new Map(reportList.map(v => [v.id, v])).values()]

    if (reportList.length === 0) {
      this.warn('no entries found, make sure the date range provided and the sources are correct')
      return
    }

    // re order by date ASC
    reportList.sort((a, b) => {
      return dayjs(a.date).diff(b.date)
    })

    // output
    let outputStr = ''
    // eslint-disable-next-line unicorn/prefer-switch
    if (validatedConfig.format === BananaOutputFileFormat.Enum.csv) {
      const fieldsColumns: Array<keyof CommonBananaReporterObj> = ['date', 'type', 'username', 'projectName', 'description']
      logger.debug('run converting json to csv')
      const csvOutput = jsonToCsv({
        fields: fieldsColumns,
        data: reportList,
      })
      outputStr = csvOutput
    } else if (validatedConfig.format === BananaOutputFileFormat.Enum.json) {
      outputStr = JSON.stringify(reportList, null, 2)
    } else if (validatedConfig.format === BananaOutputFileFormat.Enum.jsonl) {
      logger.debug('run converting json to jsonl')
      outputStr = reportList.map(r => JSON.stringify(r)).join('\n')
    }

    const fileExt = `.${validatedConfig.format}`
    let outFile = validatedConfig.out
    logger.debug(`Run.flags.out.default ${Run.flags.out.default}`)
    logger.debug(`fileExt ${fileExt}`)
    logger.debug(`outFile ${outFile}`)

    // if --out was provided and it's a directory
    // use the default file name
    try {
      if (statSync(outFile).isDirectory()) {
        outFile = resolve(outFile, Run.flags.out.default as string)
        logger.debug(`outFile is a directory, updated to ${outFile}`)
      }
    } catch (error) {
      logger.debug('statSync error', error)
    }

    // add file extension
    if (!outFile.endsWith(fileExt)) {
      if (outFile === Run.flags.out.default) {
        outFile = outFile
        .replace(`.${Run.flags.format.default}`, fileExt)
      } else {
        outFile += fileExt
      }
    }

    if (outFile.includes('$FROM') || outFile.includes('$TO')) {
      outFile = outFile
      .replace('$FROM', validatedConfig.from)
      .replace('$TO', validatedConfig.to)
    }

    logger.debug(`run saving to ${outFile} with file content of size ${outputStr.length}`)
    writeFileSync(outFile, outputStr)
    this.log(`report with ${reportList.length} entries saved to ${outFile}`)

    // this.log(`run! ${flags.config} ${JSON.stringify(configContent)}`)
  }

  async catch(error: Error): Promise<void> {
    if (error instanceof z.ZodError) {
      const errMsg = error.issues.map(i => `${i.message}: ${i.path.join(', ')}`).join('\n')
      throw new TypeError(`validation failed, ${errMsg}`)
    }

    throw error
  }
}
